StrataTest/Transform/ProcedureInlining.lean

```diff
@@ -0,0 +1,337 @@
+        return map
+    | (.cmd (.init n1 _ _e1 _), .cmd (.init n2 _ _e2 _)) =>
+      -- Omit e1 and e2 check because init may use undeclared free vars
+      -- The updateVars below must be the only place that updates the
+      -- variable name mapping.
+      IdMap.updateVars map [(n1.name,n2.name)]
+    | (.cmd (.set n1 e1 _), .cmd (.set n2 e2 _)) =>
+      if ¬ alphaEquivExprs e1 e2 map then
+        mk_err f!"RHS of sets do not match \
+        \n(subst of e1: {repr (substExpr e1 map)})\n(e2: {repr e2})"
+      else if ¬ alphaEquivIdents n1 n2 map then
+        mk_err "LHS of sets do not match"
+      else
+        return map
+    | (.cmd (.havoc n1 _), .cmd (.havoc n2 _)) =>
+      if ¬ alphaEquivIdents n1 n2 map then
+        mk_err "LHS of havocs do not match"
+      else
+        return map
+    | (.cmd (.assert _ e1 _), .cmd (.assert _ e2 _)) =>
+      if ¬ alphaEquivExprs e1 e2 map then
+        mk_err "Expressions of asserts do not match"
+      else
+        return map
+    | (.cmd (.assume _ e1 _), .cmd (.assume _ e2 _)) =>
+      if ¬ alphaEquivExprs e1 e2 map then
+        mk_err "Expressions of assumes do not match"
+      else
+        return map
+    | (_,_) =>
+      mk_err "Commands do not match"
+
+  | (_,_) => mk_err "Statements do not match"
+
+end
+
+private def alphaEquiv (p1 p2:Boogie.Procedure):Except Format Bool := do
+  if p1.body.length ≠ p2.body.length then
+    dbg_trace f!"p1: {p1}"
+    dbg_trace f!"p2: {p2}"
+    .error (s!"# statements do not match: inlined fn one has {p1.body.length}"
+        ++ s!" whereas the answer has {p2.body.length}")
+  else
+    let newmap:IdMap := IdMap.mk [] []
+    let stmts := (p1.body.zip p2.body)
+    let _ ← List.foldlM (fun (map:IdMap) (s1,s2) =>
+        alphaEquivStatement s1 s2 map)
+      newmap stmts
+    return .true
+
+
+
+def translate (t : Strata.Program) : Boogie.Program :=
+  (TransM.run Inhabited.default (translateProgram t)).fst
+
+def runInlineCall (p : Boogie.Program) : Boogie.Program :=
+  match (runProgram inlineCallStmt p .emp) with
+  | ⟨.ok res, _⟩ => res
+  | ⟨.error e, _⟩ => panic! e
+
+def checkInlining (prog : Boogie.Program) (progAns : Boogie.Program)
+    : Except Format Bool := do
+  let prog' := runInlineCall prog
+  let pp' := prog'.decls.zip progAns.decls
+  pp'.allM (fun (p,p') => do
+    match p,p' with
+    | .proc p, .proc p' =>
+      match alphaEquiv p p' with
+      | .ok _ => return .true
+      | .error msg =>
+        dbg_trace s!"{toString prog'}"
+        dbg_trace s!"{toString progAns}"
+        .error msg
+    | _, _ => .error "?")
+
+
+
+def Test1 :=
+#strata
+program Boogie;
+procedure f(x : bool) returns (y : bool) {
+  y := !x;
+};
+
+procedure h() returns () {
+  var b_in : bool;
+  var b_out : bool;
+  call b_out := f(b_in);
+};
+#end
+
+def Test1Ans :=
+#strata
+program Boogie;
+procedure f(x : bool) returns (y : bool) {
+  y := !x;
+};
+
+procedure h() returns () {
+  var b_in : bool;
+  var b_out : bool;
+  inlined: {
```

Comment:
Oh so you _are_ using a block here. Will make de Bruijn index conversion much easier.